packages/@react-spectrum/s2/src/TagGroup.tsx

```diff
@@ -86,30 +97,152 @@ const helpTextStyles = style({
   cursor: 'text'
 });
 
-function TagGroup<T extends object>(
-  {
+const InternalTagGroupContext = createContext<TagGroupProps<any>>({});
+
+function TagGroup<T extends object>(props: TagGroupProps<T>, ref: DOMRef<HTMLDivElement>) {
+  [props, ref] = useSpectrumContextProps(props, ref, TagGroupContext);
+  props = useFormProps(props);
+  let {onRemove} = props;
+  return (
+    <InternalTagGroupContext.Provider value={{onRemove}}>
+      <CollectionBuilder content={<Collection {...props} />}>
+        {collection => <TagGroupInner props={props} forwardedRef={ref} collection={collection} />}
+      </CollectionBuilder>
+    </InternalTagGroupContext.Provider>
+  );
+}
+
+/** Tags allow users to categorize content. They can represent keywords or people, and are grouped to describe an item or a search request. */
+let _TagGroup = /*#__PURE__*/ (forwardRef as forwardRefType)(TagGroup);
+export {_TagGroup as TagGroup};
+
+function TagGroupInner<T>({
+  props: {
     label,
     description,
-    items,
     labelPosition = 'top',
     labelAlign = 'start',
-    children,
     renderEmptyState,
     isEmphasized,
     isInvalid,
     errorMessage,
     UNSAFE_className = '',
     UNSAFE_style,
+    size = 'M',
     ...props
-  }: TagGroupProps<T>,
-  ref: DOMRef<HTMLDivElement>
-) {
-  [props, ref] = useSpectrumContextProps(props, ref, TagGroupContext);
+  },
+  forwardedRef: ref,
+  collection
+}: {props: TagGroupProps<T>, forwardedRef: DOMRef<HTMLDivElement>, collection: any}) {
+  let {maxRows, groupActionLabel, onGroupAction, ...otherProps} = props;
+  let {direction} = useLocale();
+  let containerRef = useRef(null);
+  let tagsRef = useRef<HTMLDivElement | null>(null);
+  let actionsRef = useRef<HTMLDivElement | null>(null);
+  let hiddenTagsRef = useRef<HTMLDivElement | null>(null);
+  let [tagState, setTagState] = useState({visibleTagCount: collection.size, showCollapseButton: false});
```

Comment:
Noticed that `tagState.visibleTagCount` is 0 if `maxRows` isn't set since the `collection.size` is 0 on first render. This means that the currently logic will render an empty ActionGroup for a TagGroup without maxRows or a row action. I'm testing setting `visibleTagCount` even if `maxRows` but maybe we can just add another check against the existance of `maxRows` to `showCollapseToggleButton`  to the TagGroup rendering